cache.go

```diff
@@ -0,0 +1,839 @@
+		secretKey: secretKey,
+		bucket:    bucket,
+		minSize:   minSize,
+		duration:  time.Duration(duration) * time.Second,
+	}
+}
+func newCacheClient(ctx *cli.Context, cfg *cacheConfig) *S3CacheClient {
+	if cfg == nil {
+		return nil
+	}
+	creds := credentials.NewStaticV4(cfg.accessKey, cfg.secretKey, "")
+	var e error
+	s3Clnt := &S3CacheClient{}
+	options := minio.Options{
+		Creds:        creds,
+		Secure:       cfg.useTLS,
+		Region:       "",
+		BucketLookup: 0,
+	}
+	u, err := url.Parse(cfg.endpoint)
+	if err != nil {
+		return nil
+	}
+	api, e := minio.NewWithOptions(u.Host, &options)
+	if e != nil {
+		return nil
+	}
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConns:          1024,
+		MaxIdleConnsPerHost:   1024,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		// Set this value so that the underlying transport round-tripper
+		// doesn't try to auto decode the body of objects with
+		// content-encoding set to `gzip`.
+		//
+		// Refer:
+		//    https://golang.org/src/net/http/transport.go?h=roundTrip#L1843
+		DisableCompression: true,
+	}
+
+	if cfg.useTLS {
+		// Keep TLS config.
+		tlsConfig := &tls.Config{
+			RootCAs: mustGetSystemCertPool(),
+			// Can't use SSLv3 because of POODLE and BEAST
+			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+			// Can't use TLSv1.1 because of RC4 cipher usage
+			MinVersion:         tls.VersionTLS12,
+			NextProtos:         []string{"http/1.1"},
+			InsecureSkipVerify: ctx.GlobalBool("insecure"),
+		}
+		tr.TLSClientConfig = tlsConfig
+	}
+
+	var transport http.RoundTripper = tr
+	// Set the new transport.
+	api.SetCustomTransport(transport)
+	// Set app info.
+	api.SetAppInfo(ctx.App.Name, ctx.App.Version)
+	// Store the new api object.
+	s3Clnt.api = api
+	cfg.endpoint = strings.TrimSuffix(cfg.endpoint, slashSeparator)
+
+	target, err := url.Parse(cfg.endpoint)
+	if err != nil {
+		console.Fatalln(fmt.Errorf("Unable to parse input arg %s: %s", cfg.endpoint, err))
+	}
+	if target.Scheme == "" {
+		target.Scheme = "http"
+	}
+	if target.Scheme != "http" && target.Scheme != "https" {
+		console.Fatalln("Unexpected scheme %s, should be http or https, please use '%s --help'",
+			cfg.endpoint, ctx.App.Name)
+	}
+	if target.Host == "" {
+		console.Fatalln(fmt.Errorf("Missing host address %s, please use '%s --help'",
+			cfg.endpoint, ctx.App.Name))
+	}
+	s3Clnt.methods = []string{http.MethodGet, http.MethodHead}
+	s3Clnt.bucket = cfg.bucket
+	s3Clnt.minSize = cfg.minSize
+	s3Clnt.healthCheckDuration = cfg.duration
+	s3Clnt.endpoint = cfg.endpoint
+	s3Clnt.useTLS = target.Scheme == "https"
+	s3Clnt.httpClient = &http.Client{Transport: tr}
+	go s3Clnt.healthCheck()
+	return s3Clnt
+}
+func sortURLParams(URL *url.URL) {
+	params := URL.Query()
```

Comment:
why do you need to sort the sortURLParams() ?

Reply:
to ensure key generated is unique, even if url params don't appear in the same order each time.